### ISSUE TITLE
CDN requests with uncommitted changes

### ISSUE THREAD

**Issue by solmsted** [yui/yogi#26]
CDN requests fail if there are any uncommitted changes in the repository.  This is a good check, but it's annoying when a repository contains multiple modules and unfinished development work for one module blocks a CDN request for another module.

Is there a way to ignore uncommitted changes and just pick a specific commit for the CDN request?

**Comment by davglass** [comment]
I'm a big fan of separating my work and never having too much in the
tree changing so I added this to encourage that. It's to keep people
from thinking they are submitting something when they are not.

I would accept a patch with an option but it's not on my list of "best
practices" :)

**Comment by davglass** [comment]
Actually now that I look at the code, there is a `--force` option that should do what you want. It looks like it even tells you that when it fails: 

https://github.com/yui/yogi/blob/master/lib/cmds/cdn.js#L66

**Comment by solmsted** [comment]
I'm sorry, I knew that, I was in a hurry and it looks like I confused myself and asked completely the wrong question.  I'll try to start over.

When I pick a specific commit to use as a CDN request, it looks like the commit hash gets sent to the gallery API but it doesn't send the actual diffs.  So when the gallery build actually pulls in the diffs from that commit, it pulls from GitHub, is this correct?

If the gallery builds are looking at the repo on GitHub, shouldn't Yogi also look there when picking a commit?  Then there wouldn't be any way for there to be uncommitted changes.  Also, as it is now, is there a chance I could submit a CDN request then fail to push to GitHub, then the gallery build would fail to pull my changes?

**Comment by davglass** [comment]
That's a good point, I suppose I "could" check Github for the list of recent commits instead of the local repo but that might also be confusing too.

Now, if I do move it to using github instead of the local tree you could "technically" submit a CDN request from anywhere and not actually have the code locally. This might be interesting..

**Comment by davglass** [comment]
You can install this from master and confirm that it does what you want it to do.

**Comment by solmsted** [comment]
I just attempted another CDN request.  It still complained that my local repo is dirty, but it also crashed when I used --force.

I got this in the terminal

```
[fedora@solmsted gallery-alea]$ yogi cdn request
yogi [info]  using yogi@0.0.49 on node@0.8.15
yogi [info]  using module: gallery-alea
yogi [info]  submitting a cdn request for gallery-alea
yogi [bail] git root is dirty, please commit first or use --force
[fedora@solmsted gallery-alea]$ yogi cdn request --force
yogi [info]  using yogi@0.0.49 on node@0.8.15
yogi [info]  using module: gallery-alea
yogi [info]  submitting a cdn request for gallery-alea
yogi [info]  fetching commits from github

/usr/local/lib/node_modules/yogi/lib/git.js:66
            res.body.forEach(function(item) {
                     ^
TypeError: Object #<Object> has no method 'forEach'
    at Request.mods.ghlog [as _callback] (/usr/local/lib/node_modules/yogi/lib/git.js:66:22)
    at Request.init.self.callback (/usr/local/lib/node_modules/yogi/node_modules/request/main.js:122:22)
    at Request.EventEmitter.emit (events.js:99:17)
    at Request.<anonymous> (/usr/local/lib/node_modules/yogi/node_modules/request/main.js:661:16)
    at Request.EventEmitter.emit (events.js:126:20)
    at IncomingMessage.Request.start.self.req.self.httpModule.request.buffer (/usr/local/lib/node_modules/yogi/node_modules/request/main.js:623:14)
    at IncomingMessage.EventEmitter.emit (events.js:126:20)
    at IncomingMessage._emitEnd (http.js:366:10)
    at HTTPParser.parserOnMessageComplete [as onMessageComplete] (http.js:149:23)
    at CleartextStream.socketOnData [as ondata] (http.js:1367:20)
```

**Comment by davglass** [comment]
K, that means the fetch from github didn't work or contained no data. Let me add an etra check to it..

**Comment by davglass** [comment]
Just added a check to the response from the github api call, if it fails then it will fall back to local.

**Comment by davglass** [comment]
I also added a config `cdn-dirty` (`yogi config set cdn-dirty true`) that will always assume `--force` on the cdn request.

**Comment by solmsted** [comment]
After the extra check was added to catch GitHub API request failures, Yogi stopped crashing on me but it always reported that the GitHub request failed.  The problem was a dumb mistake on my end, and I'm documenting the solution here in case anyone else comes across the same thing.

I had originally cloned my repo from GitHub using the HTTP address instead of the SSH address.  I fixed it by `git remote set-url origin git@github.com:correct/address/to/repo.git` then everything was better.
